### PR TITLE
DHT nodes should only handle requests on their socket

### PR DESCRIPTION
**Pull request by ssiloti**
Go through the trouble of passing the listen socket through
session_impl::on_udp_packet because otherwise we'd have to match by comparing
a tcp::endpoint with a udp::endpoint which is surprisingly expensive.

### CODE REVIEW

File: src/kademlia/node.cpp [hunk 1]
```diff
@@ -317,7 +317,8 @@ void node::incoming(msg const& m)
 			// responds to 'query' messages that it receives.
 			if (m_settings.read_only) break;
 
-			if (!native_address(m.addr)) break;
+			// only respond to requests if they're addressed to this node
```
Comment (arvidn) [inline src/kademlia/node.cpp]: is there a good reason to only filter requests (``q``) and not responses (``r``) based on which socket we received it on?

File: test/test_dht.cpp [hunk 2]
```diff
@@ -288,7 +288,7 @@ void send_dht_request(node& node, char const* msg, udp::endpoint const& ep
 	if (ec) std::printf("bdecode failed: %s\n", ec.message().c_str());
 
 	dht::msg m(decoded, ep);
-	node.incoming(m);
+	node.incoming(node.m_sock, m);
```
Comment (arvidn) [inline test/test_dht.cpp]: is this the only reason ``m_sock`` was made ``public``?

File: include/libtorrent/kademlia/node.hpp [hunk 3]
```diff
@@ -241,6 +241,7 @@ class TORRENT_EXTRA_EXPORT node
 public:
 	routing_table m_table;
 	rpc_manager m_rpc;
+	aux::listen_socket_handle const m_sock;
```
Comment (arvidn) [inline include/libtorrent/kademlia/node.hpp]: I can't find a strong reason to make this public. perhaps I'm missing something.

File: include/libtorrent/kademlia/node.hpp [hunk 4]
```diff
@@ -253,6 +254,8 @@ class TORRENT_EXTRA_EXPORT node
 
 	static protocol_descriptor const& map_protocol_to_descriptor(udp protocol);
 
+	socket_manager* m_sock_man;
```
Comment (arvidn) [inline include/libtorrent/kademlia/node.hpp]: did this move just to order it among other pointers?

**Comment by arvidn** [comment]
would it be a lot of trouble to set up a test that exercise this? a unit test might be simplest

File: src/kademlia/node.cpp (hunk 1 above)
Comment (ssiloti) [reply src/kademlia/node.cpp]: Nodes can send requests using a listen socket other than their own for cross-network bootstrap. They need to be able to receive the response.

File: test/test_dht.cpp (hunk 2 above)
Comment (ssiloti) [reply test/test_dht.cpp]: Yes. Having made `m_sock` `const` I didn't see any harm in doing so.

File: include/libtorrent/kademlia/node.hpp (hunk 4 above)
Comment (ssiloti) [reply include/libtorrent/kademlia/node.hpp]: I moved it up to keep it next to `m_sock` because they usually get accessed together.

**Pull request merged**